File: rust_Solutions/code_11.rs
/*
//  Question: Merge two sorted arrays in Rust

// Observations: 
    * As we have sorted array given, we just have to merge it & from a new array...

// Intrusion: 
    * We can use two pointer approach: 
        * first pointer in first array
        * Second pointer in second array..
    * We will compare each pointer element & whose is smaller, we will store in third array..
    * Last we have to look after the left over element, that are not stores..
        * we can check this in both the array...
    * and last we will print the final array..

// Example TestCases: 
// Input:
4 5
10 15 20 40
5 6 6 10 15

// Output: 
5 6 6 10 10 15 15 20 40 


// Time Complexity: 
    O(n + m), sum of size of first array & second array..


*/


// Merge two sorted arrays
fn merge(a: &[i32], b: &[i32]) -> Vec<i32> {
    let mut i = 0;
    let mut j = 0;
    let mut k = 0;
    let n = a.len();
    let m = b.len();
    let mut arr = vec![0; n + m];

    while i < n && j < m {
        if a[i] < b[j] {
            arr[k] = a[i];
            i += 1;
        } else {
            arr[k] = b[j];
            j += 1;
        }
        k += 1;
    }

    while i < n {
        arr[k] = a[i];
        i += 1;
        k += 1;
    }

    while j < m {
        arr[k] = b[j];
        j += 1;
        k += 1;
    }

    arr
}

fn main() {
    let a = [10, 15, 20, 40];
    let b = [5, 6, 6, 10, 15];
    
    let merged_array = merge(&a, &b);
    
    for num in &merged_array {
        print!("{} ", num);
    }
    println!();
}

File: cpp_solutions/code_11.cpp
#include<bits/stdc++.h>
using namespace std;

// Merge two sorted array:
void merge(int a[], int b[], int n, int m){
    int i=0,j=0;    // using two pointers that help to store array in sorted order in third array.
    int arr[n+m];   // initialization of third array of size (n+m), which is of both the array size..
    int k = 0;  // pointer for third array that will store element from both the array..
    // when pointer i < n & j < m, then only it goes into the scope..
    while(i<n && j<m){
        // condition when first array element is smaller than the second array element..
        if(a[i] < b[j]){
            arr[k++] = a[i++];  // storing the element & increment the pointer..
        }
        // condition when second array element is lesser than first array element..
        else{
            arr[k++] = b[j++];  // storing & incrementing the pointers..
        }
    }
    // for left over in first array...
    while(i<n){
        arr[k++] = a[i++];  // storing the first array element to the final array..
    }
    // for left over in second array...
    while(j<m){
        arr[k++] = b[j++];  // storing the second array element to the final array..
    }

    // last we print the third array..
    for(int i = 0;i<m+n;i++){
        cout << arr[i] << " ";
    }
    cout << endl;
}


int main(){
    int n, m;
    cin >> n >> m;
    int a[n];
    int b[m];
    for(int i=0;i<n;i++) cin >> a[i];
    for(int i=0;i<m;i++) cin >> b[i];
	
	merge(a,b,n,m);

    return 0;
}